FAD-5629 Suppressions - Upload List

bkemper opened the pull request:
_Refer to [FAD-5629](https://jira.int.messagesystems.com/browse/FAD-5629) for AC._  Note: entering a single entry is not included and will be addressed in a later ticket.

**Before**

<img width="1385" alt="screen shot 2018-01-25 at 3 58 45 pm" src="https://user-images.githubusercontent.com/1335605/35411950-a9aaba9a-01e8-11e8-91d3-f2cbb34837ea.png">

**After**

<img width="1378" alt="screen shot 2018-02-07 at 1 34 27 am" src="https://user-images.githubusercontent.com/1335605/35901906-1d5f2620-0ba7-11e8-8be0-4137816b9057.png">

**How to test?**
- Use these [examples.zip](https://github.com/SparkPost/2web2ui/files/1702002/Archive.zip) files to test different states.
- Quick sanity check that recipient list upload still works as expected.

rajumsys commented on src/pages/suppressions/CreatePage.js:
```diff
@@ -0,0 +1,16 @@
+  render() {
+    return (
+      <Page title="Create Suppression List">
+        Hi
```
hello :)

bkemper commented:
Rebased and ready for review.

ewandennis commented:
Visual review first:
 - Uploading `example-wrong-type.csv` [threw on an unhandled rejection](https://sentry.io/sparkpost/2web2ui-dev/issues/452242423/).
 - I'm not sure about the 'only for a single subaccount` checkbox.
    - Just showing the subaccount typeahead with a note saying "leave this blank to upload to the master account suppression list" might work.
    - If we continue to hide the typeahead, maybe a radio group would make both options explicit:
        - Upload to the master account
        - Upload to a subaccount
        - ^^ imperative language is also good here IMO

ewandennis commented on src/pages/suppressions/tests/CreatePage.test.js:
```diff
@@ -0,0 +1,63 @@
+  }
+});
+
+describe('CreatePage.handleSubmit', async() => {
```
You're not testing that the correct action creators were called here. which leaves the non-subaccount case untested too. Any reason?

bkemper replied:
No real reason.  I just glazed over it after setting up the mocks.  Side note, I miss RSpec's mock with argument constraints `allow(obj).to receive(:message).with('an argument').and_return(...)`.

bkemper replied:
Addressed... e97c1592

ewandennis commented on src/pages/suppressions/CreatePage.js:
```diff
@@ -0,0 +1,104 @@
+              status="danger"
+            />
+          )}
+          <Panel title="Upload a List of Suppressions">
```
Do we need the page and panel titles? I might just pick one.

bkemper replied:
I think we need a page title to keep a uniformed look.  How does this look without the panel title?  

<img width="990" alt="screen shot 2018-02-07 at 11 48 39 am" src="https://user-images.githubusercontent.com/1335605/35929201-edb19056-0bfc-11e8-9638-8d5ce9a0de8b.png">

bkemper replied:
Addressed... b0715041

bkemper commented:
Need to do some more transformations and validation before sending request to match webui.

- confirm required fields are present (recipient and type)
- typo proof the type field
- regex test recipient field
- support for deprecated fields (transactional and non_transactional ~> type, email ~> recipient)

bkemper commented on src/actions/suppressions.js:
```diff
@@ -115,3 +116,35 @@ export function deleteSuppression(suppression) {
   });
 }
 
+// SEE: https://github.com/SparkPost/sparkpost-api-documentation/blob/master/services/suppression-list.md#insert-or-update-list-entries-put
```
Addressed... 079ec707

bkemper commented on src/pages/suppressions/CreatePage.js:
```diff
@@ -0,0 +1,112 @@
+        <form onSubmit={reduxFormSubmit(this.handleSubmit)}>
+          {!submitting && submitFailed && (
+            <ApiErrorBanner
+              errorDetails={this.errorDetails}
```
Is 50 a reasonable max?

bkemper replied:
Addressed... f3a8dbef

bkemper commented on src/pages/suppressions/CreatePage.js:
```diff
@@ -0,0 +1,112 @@
+                component={FileFieldWrapper}
+                disabled={submitting}
+                name="suppressionsFile"
+                label="Comma Separated Values (CSV) File of Suppressions"
```
Addressed... b0715041

bkemper commented on src/helpers/validation.js:
```diff
@@ -34,6 +34,15 @@ export function endsWithWhitespace(value) {
   return /[\S]$/.test(value) ? undefined : 'Can\'t end in a whitespace character';
 }
 
+export function fileWithContents(file) {
```
Addressed... 3060911b

bkemper commented on src/pages/suppressions/CreatePage.js:
```diff
@@ -0,0 +1,112 @@
+
+    return parseError
+      ? parseError.details || parseError.message
+      : _.get(persistError, 'response.data.errors[0].message', persistError.message);
```
Addressed... 90cf2e17

bkemper commented:
@jasonrhodes This is ready for your final review.  Please let me know if you want to do a final walkthrough.

jasonrhodes commented on src/actions/suppressions.js:
```diff
@@ -116,9 +116,30 @@ export function deleteSuppression(suppression) {
+    if (type && !LIKE_NON.test(type)) { type = 'transactional'; }
+
+    // Convert deprecated recipient fields
+    if (email) { recipient = email; }
```
`if (email && !recipient)` ?

bkemper replied:
Addressed... 34f03d30

jasonrhodes commented on src/actions/suppressions.js:
```diff
@@ -116,9 +116,30 @@ export function deleteSuppression(suppression) {
+    if (email) { recipient = email; }
+
+    // Trim whitespace from recipient email
+    if (recipient) { recipient = _.trim(recipient); } // for FAD-5095
```
is `recipient` guaranteed at this point?

bkemper replied:
Yes.  I'll remove that condition.

bkemper replied:
Addressed... a8de3ad2

jasonrhodes commented on src/actions/tests/suppressions.test.js:
```diff
@@ -66,6 +66,32 @@ describe('Action Creator: Suppressions', () => {
+      expect(action).toMatchSnapshot();
+    });
+
+    it('with typo in type field', () => {
```
maybe "with non-standard type fields, parses based on regex" or something? had trouble remembering what this was testing at first glance

bkemper replied:
Addressed... 00eee30a

jasonrhodes commented on src/actions/tests/suppressions.test.js:
```diff
@@ -66,6 +66,32 @@ describe('Action Creator: Suppressions', () => {
+      ]);
+
+      expect(action).toMatchSnapshot();
+    });
```
Can we add a test where there is both `type` _and_ `transactional` `non_transactional` keys, to prove that type has precedence?

bkemper replied:
Addressed... dde53c89

jasonrhodes commented on src/actions/tests/suppressions.test.js:
```diff
@@ -66,6 +66,32 @@ describe('Action Creator: Suppressions', () => {
+      ]);
+
+      expect(action).toMatchSnapshot();
+    });
```
Add a test where there is both `recipient` and `email` keys to prove precedence?

jasonrhodes commented on src/actions/helpers/localFileParseRequest.js:
```diff
@@ -1,6 +1,18 @@
+  }
+};
+
+export const hasField = (...expectedFields) => ({ meta: { fields }}) => {
```
Agree on renaming this, maybe `hasAtLeastOneField` lol idk

jasonrhodes commented:
@bkemper a few general thoughts and questions:
* does papaparse do anything but CSV parsing? if not, should that action flow be named around CSV instead of more generic local file parse?
* for my own learning, what's the advantage of using a getter over a method?
* looks like you are calling meta.validate as if it is a single function but it's actually an array of functions, so that's blowing up

jasonrhodes commented on src/actions/helpers/localFileParseRequest.js:
```diff
@@ -23,10 +38,21 @@ export default ({ meta, type, parser = Papa }) => async(dispatch) => {
       }
+
+      // Custom validation of data
+      if (validate && (message = validate({ data, meta }))) {
```
Here's the bug ^

Should be able to fix by reducing over the validate array or something like that?

bkemper commented:
> does papaparse do anything but CSV parsing? if not, should that action flow be named around CSV instead of more generic local file parse?

No, papaparse only handles CSV files.  Addressed... 10c87187

> for my own learning, what's the advantage of using a getter over a method?

Oops.  I forgot to switch it back after refactoring it.  Addressed... a2f787b2

> looks like you are calling meta.validate as if it is a single function but it's actually an array of functions, so that's blowing up

Yikes!  Addressed... 461a48c0

jasonrhodes commented on src/actions/helpers/csvFileParseRequest.js:
```diff
@@ -41,8 +41,10 @@ export default ({ meta: { file, validate }, type, parser = Papa }) => async(disp
-      // Custom validation of data
-      if (validate && (message = validate({ data, meta }))) {
+      // Return the first validation error message
+      const message = _.find(validate, (v) => v({ data, meta }));
```
oh good idea :clap:

jasonrhodes commented on src/actions/helpers/csvFileParseRequest.js:
```diff
@@ -0,0 +1,66 @@
+      }
+
+      // Return the first validation error message
+      const message = _.find(validate, (v) => v({ data, meta }));
```
I thought this was an awesome idea until I realized it didn't quite work, doh. Haha.

See #282

bkemper replied:
Double yikes!  I should have seen this in the snapshots, `"message": [Function],`.
